### PR TITLE
📚 Accumulate completions for logging

### PR DESCRIPTION
**Pull request by lewtun**
# What does this PR do?

This PR fixes a bug where the `rich` and `wandb` tables were not logging the complete set of unique prompts for `gradient_accumulation_steps > 1`. Previously what happened is that the tables we logged for each gradient accumulation step, which is confusing when inspecting the logs for unique prompts.

Here's two runs with/without the fix for:

* `main`: https://wandb.ai/huggingface/huggingface/runs/j3vwwb5t/workspace?nw=nwuserlewtun
* `this_pr`: https://wandb.ai/huggingface/huggingface/runs/78q63hev/overview

These runs have:

* `gradient_accumulation_steps=4`
* `num_generations=16`
* `per_device_train_batch_size=4`
* `num_gpus = 8`

so we expect `num_unique_prompts=8`:

```
num_unique_prompts = num_gpus * per_device_train_batch_size * gradient_accumulation_steps / num_generations = 8
```

We see that without the fix, `wandb` only displays 2 unique prompts in the table (i.e. once per gradient accumulation step). With the fix, the prompts are grouped together per step, which incidentally makes the table scrolling faster in the UI as we have fewer steps to query.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/grpo_config.py [hunk 1]
```diff
@@ -136,6 +136,9 @@ class GRPOConfig(TrainingArguments):
             installed, it prints the sample. If `wandb` logging is enabled, it logs it to `wandb`.
         num_completions_to_print (`int` or `None`, *optional*, defaults to `None`):
             Number of completions to print with `rich`. If `None`, all completions are logged.
+        wandb_log_unique_prompts (`bool`, *optional*, defaults to `False`):
```
Comment (lewtun) [inline trl/trainer/grpo_config.py:139]: Missed from #3191

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -904,36 +904,63 @@ def _generate_and_score_completions(
...
+        is_last_step_in_grad_accum = (
+            self._step % self.args.gradient_accumulation_steps == self.args.gradient_accumulation_steps - 1
+        )
+        should_log_completions = (
+            self.log_completions
+            and self.state.global_step % self.args.logging_steps == 0
+            and is_last_step_in_grad_accum
+        )
+
+        # Collect data for logging throughout the accumulation steps
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:918]: I think `self.control.should_log` achieves the same goal. Let me check

Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:918]: So yes it achieves the same goal but this flag is only update after the `loss.backward` so we can't use it here. I'm starting to think that it would make more sense to have all this code in `self.log`. I'm checking if it's possible.

Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:918]: So yes It's possible, I'll do it in a follow-up PR

**Review by qgallouedec — APPROVED** [review]
